### ISSUE TITLE
[C] Intel Edison - Unable To Receives Messages Continuously

### ISSUE THREAD

**Issue by cnvery** [Azure/azure-iot-sdks#757]
Hi all,

I faced one weird problem on Intel Edison.

I am trying to change the sample code from _simplesample_amqp.c_.

It is able to send and receives messages first time without any problem. But, not for second times and so on.

I defined one model **ContosoAnemometer** with 9 data and 1 action.

During the first time, Intel Edison is able to transmit and receives JSON data from Microsoft Azure IoT cloud server. The function _SetGreenTime_ (Important I/O function inside) is ok to runs without any issues.

**Below are JSON format for received data:**
SetGreenTime
{
int north_green_time,
int north_priority, 
int east_green_time, 
int east_priority, 
int south_green_time, 
int south_priority, 
int west_green_time, 
int west_priority
}

Example for first received JSON data from cloud server:
north_green_time = 10
north_priority = 1
east_green_time = 5
east_priority = 3
south_green_time = 8
south_priority = 2
west_green_time = 2
west_priority = 4

Second time, Intel Edison send JSON data successfully without any issues. 

However, the function _SetGreenTime_ will read back first received JSON data instead of second received JSON data. (Despite cloud server is sending second JSON data correctly)

Example for second received JSON data from cloud server:
north_green_time = 13
north_priority = 1
east_green_time = 6
east_priority = 3
south_green_time = 9
south_priority = 2
west_green_time = 1
west_priority = 4

Due to this, it make my remaining system cannot works.

Can anyone tell which part of my code went wrong? Since it is working fine for first time.

Is the first received JSON data will store in C file forever, and cannot be altered later on? (Well, this is troublesome for users wish to receives different data continuously)

Thanks!

**Attached <Main.c>:**
```
#include stdio.h
#include stdlib.h
#include stdint.h
#include mraa.h
#include "Application.h"

// Receive JSON Messages Library
#include string.h
#include "jsmn.h"

// Microsoft Azure IoT Library
#include "serializer.h"
#include "azure_c_shared_utility/threadapi.h"
#include "azure_c_shared_utility/sastoken.h"
#include "iothub_client.h"
#include "iothubtransportamqp.h"
#include "iothub_client_ll.h"
#include "azure_c_shared_utility/platform.h"

#ifdef MBED_BUILD_TIMESTAMP
#include "certs.h"
#endif // MBED_BUILD_TIMESTAMP

/*String containing Hostname, Device Id & Device Key in the format:             */
/*  "HostName=<host_name>;DeviceId=<device_id>;SharedAccessKey=<device_key>"    */
static const char* connectionString = "*******************************";

// Define the Model
BEGIN_NAMESPACE(WeatherStation);

DECLARE_MODEL(ContosoAnemometer,
WITH_DATA(ascii_char_ptr, DeviceId),
WITH_DATA(int, nt1),
WITH_DATA(int, et1),
WITH_DATA(int, st1),
WITH_DATA(int, wt1),
WITH_DATA(int, nt2),
WITH_DATA(int, et2),
WITH_DATA(int, st2),
WITH_DATA(int, wt2),
WITH_ACTION(SetGreenTime, int, north_green_time, int, north_priority, int, east_green_time, int, east_priority, int, south_green_time, int, south_priority, int, west_green_time, int, west_priority)
);

END_NAMESPACE(WeatherStation);

DEFINE_ENUM_STRINGS(IOTHUB_CLIENT_CONFIRMATION_RESULT, IOTHUB_CLIENT_CONFIRMATION_RESULT_VALUES)
 
void sendCallback(IOTHUB_CLIENT_CONFIRMATION_RESULT result, void* userContextCallback)
{
    int messageTrackingId = (intptr_t)userContextCallback;

    (void)printf("Message Id: %d Received.\r\n", messageTrackingId);
 
    (void)printf("Result Call Back Called! Result is: %s \r\n", ENUM_TO_STRING(IOTHUB_CLIENT_CONFIRMATION_RESULT, result));
}

static void sendMessage(IOTHUB_CLIENT_HANDLE iotHubClientHandle, const unsigned char* buffer, size_t size)
{
    static unsigned int messageTrackingId;
    IOTHUB_MESSAGE_HANDLE messageHandle = IoTHubMessage_CreateFromByteArray(buffer, size);
    if (messageHandle == NULL)
    {
        printf("unable to create a new IoTHubMessage\r\n");
    }
    else
    {
        if (IoTHubClient_SendEventAsync(iotHubClientHandle, messageHandle, sendCallback, (void*)(uintptr_t)messageTrackingId) != IOTHUB_CLIENT_OK)
        {
            printf("failed to hand over the message to IoTHubClient");
        }
        else
        {
            printf("IoTHubClient accepted the message for delivery\r\n");
        }

        IoTHubMessage_Destroy(messageHandle);
    }
    free((void*)buffer);
    messageTrackingId++;
}

static IOTHUBMESSAGE_DISPOSITION_RESULT IoTHubMessage(IOTHUB_MESSAGE_HANDLE message, void* userContextCallback)
{
    IOTHUBMESSAGE_DISPOSITION_RESULT result;
    const unsigned char* buffer;
    size_t size;
    if (IoTHubMessage_GetByteArray(message, &buffer, &size) != IOTHUB_MESSAGE_OK)
    {
        printf("unable to IoTHubMessage_GetByteArray\r\n");
        result = EXECUTE_COMMAND_ERROR;
    }

    else
    {
	int i;
	int r;
	jsmn_parser p;
	jsmntok_t t[128];

	jsmn_init(&p);
	r = jsmn_parse(&p, buffer, strlen(buffer), t, sizeof(t) / sizeof(t[0]));

	char keyString[129];
 
	for(i = 1;i < r;i++)
	{
		jsmntok_t key = t[i];
		unsigned int length = key.end - key.start;
		memcpy(keyString, &buffer[key.start], length);
		keyString[length] = '\0';
		printf("Key: %s\n", keyString);
	}
 
        /*buffer is not zero terminated*/
        char* temp = malloc(size + 1);
        if (temp == NULL)
        {
            printf("failed to malloc\r\n");
            result = EXECUTE_COMMAND_ERROR;
        }
        else
        {
            memcpy(temp, buffer, size);
            temp[size] = '\0';
            EXECUTE_COMMAND_RESULT executeCommandResult = EXECUTE_COMMAND(userContextCallback, temp);
            result = (executeCommandResult == EXECUTE_COMMAND_ERROR) ? IOTHUBMESSAGE_ABANDONED : (executeCommandResult == EXECUTE_COMMAND_SUCCESS) ? IOTHUBMESSAGE_ACCEPTED : IOTHUBMESSAGE_REJECTED;
            free(temp);
	    (void)printf("Received Messages Success!!\r\n");
	    delay_seconds(2);
	    SendMessage2();					// Send Messages To Microsoft Azure Cloud Server Again
	    System_Wait();
        }
    }
    return result;
}

void SendMessage1(void)
{
    if (platform_init() != 0)
    {
        printf("Failed to initialize the platform.\r\n");
    }

    else
    {
        if (serializer_init(NULL) != SERIALIZER_OK)
        {
            (void)printf("Failed on serializer_init\r\n");
        }

        else
        {
            /* Setup IoTHub client configuration */
            IOTHUB_CLIENT_HANDLE iotHubClientHandle = IoTHubClient_CreateFromConnectionString(connectionString, AMQP_Protocol);
            srand((unsigned int)time(NULL));

            if (iotHubClientHandle == NULL)
            {
                (void)printf("Failed on IoTHubClient_Create\r\n");
            }

            else
            {
// #ifdef MBED_BUILD_TIMESTAMP
                // For mbed add the certificate information
                if (IoTHubClient_SetOption(iotHubClientHandle, "TrustedCerts", certificates) != IOTHUB_CLIENT_OK)
                {
                    (void)printf("failure to set option \"TrustedCerts\"\r\n");
                }
// #endif // MBED_BUILD_TIMESTAMP

                ContosoAnemometer* myWeather = CREATE_MODEL_INSTANCE(WeatherStation, ContosoAnemometer);

                if (myWeather == NULL)
                {
                    (void)printf("Failed on CREATE_MODEL_INSTANCE\r\n");
                }

                else
                {
                    unsigned char* destination;
                    size_t destinationSize;

                    if (IoTHubClient_SetMessageCallback(iotHubClientHandle, IoTHubMessage, myWeather) != IOTHUB_CLIENT_OK)
                    {
                        printf("unable to IoTHubClient_SetMessageCallback\r\n");
                    }

                    else
                    {
                        myWeather->DeviceId = "Edison01";
                        myWeather->nt1 = rand() % 50;
                        myWeather->et1 = rand() % 50;
                        myWeather->st1 = rand() % 50;
                        myWeather->wt1 = rand() % 50;
                        myWeather->nt2 = rand() % 50;
                        myWeather->et2 = rand() % 50;
                        myWeather->st2 = rand() % 50;
                        myWeather->wt2 = rand() % 50;

                        if (SERIALIZE(&destination, &destinationSize, myWeather->DeviceId, myWeather->nt1, myWeather->et1, myWeather->st1, myWeather->wt1, myWeather->nt2, myWeather->et2, myWeather->st2, myWeather->wt2) != IOT_AGENT_OK)
                        {
                            (void)printf("Failed to serialize\r\n");
                        }

                        else
                        {
                            sendMessage(iotHubClientHandle, destination, destinationSize);
                            (void)printf("Serialize Success!!\r\n");
                        }

                        /* wait for commands */
                        (void)getchar();
                    }
                    DESTROY_MODEL_INSTANCE(myWeather);
                }
                IoTHubClient_Destroy(iotHubClientHandle);
            }
            serializer_deinit();
        }
        platform_deinit();
    }
}

void SendMessage2(void)
{
	/* Setup IoTHub client configuration */
        IOTHUB_CLIENT_HANDLE iotHubClientHandle = IoTHubClient_CreateFromConnectionString(connectionString, AMQP_Protocol);
        srand((unsigned int)time(NULL));

        if (iotHubClientHandle == NULL)
        {
            (void)printf("Failed on IoTHubClient_Create\r\n");
        }

        else
        {
// #ifdef MBED_BUILD_TIMESTAMP
            // For mbed add the certificate information
            if (IoTHubClient_SetOption(iotHubClientHandle, "TrustedCerts", certificates) != IOTHUB_CLIENT_OK)
            {
                (void)printf("failure to set option \"TrustedCerts\"\r\n");
            }
// #endif // MBED_BUILD_TIMESTAMP

            ContosoAnemometer* myWeather = CREATE_MODEL_INSTANCE(WeatherStation, ContosoAnemometer);
 
            if (myWeather == NULL)
            {
                (void)printf("Failed on CREATE_MODEL_INSTANCE\r\n");
            }

            else
            {
                unsigned char* destination;
                size_t destinationSize;

                if (IoTHubClient_SetMessageCallback(iotHubClientHandle, IoTHubMessage, myWeather) != IOTHUB_CLIENT_OK)
                {
                    printf("unable to IoTHubClient_SetMessageCallback\r\n");
                }

                else
                {
                    myWeather->DeviceId = "Edison01";
                    myWeather->nt1 = rand() % 50;

                    myWeather->et1 = rand() % 50;
                    myWeather->st1 = rand() % 50;
                    myWeather->wt1 = rand() % 50;
                    myWeather->nt2 = rand() % 50;
                    myWeather->et2 = rand() % 50;
                    myWeather->st2 = rand() % 50;
                    myWeather->wt2 = rand() % 50;
 
                    if (SERIALIZE(&destination, &destinationSize, myWeather->DeviceId, myWeather->nt1, myWeather->et1, myWeather->st1, myWeather->wt1, myWeather->nt2, myWeather->et2, myWeather->st2, myWeather->wt2) != IOT_AGENT_OK)
                    {
                        (void)printf("Failed to serialize\r\n");
                    }

                    else
                    {
                        sendMessage(iotHubClientHandle, destination, destinationSize);
                        (void)printf("Serialize Success!!\r\n");
                    }
 
                    /* wait for commands */
                    (void)getchar();
                }
                DESTROY_MODEL_INSTANCE(myWeather);
            }
            IoTHubClient_Destroy(iotHubClientHandle);
        }
}

EXECUTE_COMMAND_RESULT SetGreenTime(ContosoAnemometer* device, int north_green_time, int north_priority, int east_green_time, int east_priority, int south_green_time, int south_priority, int west_green_time, int west_priority)
{
    direction temp_junction;

    (void)device;
    (void)printf("North Green Time is: %d.\r\n", north_green_time);
    (void)printf("North Priority is: %d.\r\n", north_priority);
    (void)printf("East Green Time is: %d.\r\n", east_green_time);
    (void)printf("East Priority is: %d.\r\n", east_priority);
    (void)printf("South Green Time is: %d.\r\n", south_green_time);
    (void)printf("South Priority is: %d.\r\n", south_priority);
    (void)printf("West Green Time is: %d.\r\n", west_green_time);
    (void)printf("West Priority is: %d.\r\n", west_priority);

    temp_junction.north_green_time = north_green_time;
    temp_junction.north_priority = north_priority;
    temp_junction.east_green_time = east_green_time;
    temp_junction.east_priority = east_priority;
    temp_junction.south_green_time = south_green_time;
    temp_junction.south_priority = south_priority;
    temp_junction.west_green_time = west_green_time;
    temp_junction.west_priority = west_priority;

    System_Run2(temp_junction);
    SendMessage2();						// Send Messages To Microsoft Azure Cloud Server Again

    return EXECUTE_COMMAND_SUCCESS;
}
 
void System_Run(void)
{
	// System Running Code Here - Only Related I/O Stuffs.............
}
 
void System_Run2(direction junction)
{
	// System Running Code Here - Only Related I/O Stuffs.............
}


int main(int argc, char** argv)
{
   SendMessage1();
}
```

**Comment by jtanner-msft** [comment]
Hi @cnvery,

  I saw your post about having some problems so have started trying to investigate this issue for you.

  First, since it’s been a little while I wanted to check if you were still seeing this problem, or if you had determined a solution already?

  The other thing I wanted to check on is if you’ve tried to test this code on either Windows, or a desktop version of Linux like Ubuntu?  If so, did you continue to see the issue there, or is it only specific to when you run this sample on the Intel Edison?

  So far I’ve been working to get the current version of simplesample_amqp.c that you had referenced to work as-is on the Edison.  Did your version work okay for you?  If not I have some changes that I can suggest around the certificates that you might need to know about.

Thanks,

**Comment by jtanner-msft** [comment]
Hi @cnvery,
 
Since it's been a week now and I haven't seen any response from you on this issue I'm going to proceed with closing it at this time.  If you do still require assistance please feel free to re-open this issue or submit a new one referencing it in the future.

Thanks,

**Comment by cnvery** [comment]
Hi @jtanner-msft,

The problem still remain, and I dont have privilege to re-open this issue.

I have tried the code on Windows Visual Studio 2015, the output is same as Intel Edison.

It ables to send and receives first JSON data without any issues.

For second JSON data, it able to send second JSON data to cloud without any issues. However, it receives back first JSON data instead of second JSON data.

Kindly provide a solution for this.

Thank you.

**Comment by jtanner-msft** [comment]
Hi Cnvery,

  I saw in your code that I didn’t have some of the libraries that you were using, such as jsmntok, etc.

  But in reviewing the code I believe I understood the core of what you were trying to accomplish with your modifications.  I have changed the simplesample_amqp.c file so it now looks like this:

```
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

#include <stdlib.h>

#include <stdio.h>
#include <stdint.h>

#include "serializer.h"
#include "azure_c_shared_utility/threadapi.h"
#include "azure_c_shared_utility/sastoken.h"
#include "azure_c_shared_utility/platform.h"
#include "iothub_client.h"
#include "iothubtransportamqp.h"
#include "iothub_client_ll.h"

#ifdef MBED_BUILD_TIMESTAMP
#include "certs.h"
#endif // MBED_BUILD_TIMESTAMP

/*String containing Hostname, Device Id & Device Key in the format:             */
/*  "HostName=<host_name>;DeviceId=<device_id>;SharedAccessKey=<device_key>"    */
static const char* connectionString = "HostName=jttest33.azure-devices.net;DeviceId=WinAMQP;SharedAccessKey=q4oDaD5MD292oc5nGpy268OXVXCDTNoZJnEBkrpcLOo=";


// Define the Model
BEGIN_NAMESPACE(WeatherStation);

DECLARE_MODEL(ContosoAnemometer,
WITH_DATA(ascii_char_ptr, DeviceId),
WITH_DATA(int, nt1),
WITH_DATA(int, et1),
WITH_DATA(int, st1),
WITH_DATA(int, wt1),
WITH_DATA(int, nt2),
WITH_DATA(int, et2),
WITH_DATA(int, st2),
WITH_DATA(int, wt2),
WITH_ACTION(SetGreenTime, int, north_green_time, int, north_priority, int, east_green_time, int, east_priority, int, south_green_time, int, south_priority, int, west_green_time, int, west_priority)
);

END_NAMESPACE(WeatherStation);


IOTHUB_CLIENT_HANDLE iotHubClientHandle = NULL;
ContosoAnemometer* myWeather = NULL;

void SendMessage2(void);

EXECUTE_COMMAND_RESULT SetGreenTime(ContosoAnemometer* device, int north_green_time, int north_priority, int east_green_time, int east_priority, int south_green_time, int south_priority, int west_green_time, int west_priority)
{

	(void)device;
	(void)printf("North Green Time is: %d.\r\n", north_green_time);
	(void)printf("North Priority is: %d.\r\n", north_priority);
	(void)printf("East Green Time is: %d.\r\n", east_green_time);
	(void)printf("East Priority is: %d.\r\n", east_priority);
	(void)printf("South Green Time is: %d.\r\n", south_green_time);
	(void)printf("South Priority is: %d.\r\n", south_priority);
	(void)printf("West Green Time is: %d.\r\n", west_green_time);
	(void)printf("West Priority is: %d.\r\n", west_priority);

	myWeather->nt1 = north_green_time;
	myWeather->et1 = east_green_time;
	myWeather->st1 = south_green_time;
	myWeather->wt1 = west_green_time;
	myWeather->nt2 = north_priority;
	myWeather->et2 = east_priority;
	myWeather->st2 = south_priority;
	myWeather->wt2 = west_priority;


	SendMessage2();                       // Send Messages To Microsoft Azure Cloud Server Again

	return EXECUTE_COMMAND_SUCCESS;
}




void sendCallback(IOTHUB_CLIENT_CONFIRMATION_RESULT result, void* userContextCallback)
{
    unsigned int messageTrackingId = (unsigned int)(uintptr_t)userContextCallback;

    (void)printf("Message Id: %u Received.\r\n", messageTrackingId);

    (void)printf("Result Call Back Called! Result is: %s \r\n", ENUM_TO_STRING(IOTHUB_CLIENT_CONFIRMATION_RESULT, result));
}


static void sendMessage(IOTHUB_CLIENT_HANDLE iotHubClientHandle, const unsigned char* buffer, size_t size)
{
    static unsigned int messageTrackingId;
    IOTHUB_MESSAGE_HANDLE messageHandle = IoTHubMessage_CreateFromByteArray(buffer, size);
    if (messageHandle == NULL)
    {
        printf("unable to create a new IoTHubMessage\r\n");
    }
    else
    {
        if (IoTHubClient_SendEventAsync(iotHubClientHandle, messageHandle, sendCallback, (void*)(uintptr_t)messageTrackingId) != IOTHUB_CLIENT_OK)
        {
            printf("failed to hand over the message to IoTHubClient");
        }
        else
        {
            printf("IoTHubClient accepted the message for delivery\r\n");
        }

        IoTHubMessage_Destroy(messageHandle);
    }
    free((void*)buffer);
    messageTrackingId++;
}

static IOTHUBMESSAGE_DISPOSITION_RESULT IoTHubMessage(IOTHUB_MESSAGE_HANDLE message, void* userContextCallback)
{
    IOTHUBMESSAGE_DISPOSITION_RESULT result;
    const unsigned char* buffer;
    size_t size;
    if (IoTHubMessage_GetByteArray(message, &buffer, &size) != IOTHUB_MESSAGE_OK)
    {
        printf("unable to IoTHubMessage_GetByteArray\r\n");
        result = EXECUTE_COMMAND_ERROR;
    }
    else
    {
        /*buffer is not zero terminated*/
        char* temp = malloc(size + 1);
        if (temp == NULL)
        {
            printf("failed to malloc\r\n");
            result = EXECUTE_COMMAND_ERROR;
        }
        else
        {
            memcpy(temp, buffer, size);
            temp[size] = '\0';
            EXECUTE_COMMAND_RESULT executeCommandResult = EXECUTE_COMMAND(userContextCallback, temp);
            result =
                (executeCommandResult == EXECUTE_COMMAND_ERROR) ? IOTHUBMESSAGE_ABANDONED :
                (executeCommandResult == EXECUTE_COMMAND_SUCCESS) ? IOTHUBMESSAGE_ACCEPTED :
                IOTHUBMESSAGE_REJECTED;
            free(temp);
        }
    }
    return result;
}

void simplesample_amqp_run(void)
{
    if (platform_init() != 0)
    {
        printf("Failed to initialize the platform.\r\n");
    }
    else
    {
        if (serializer_init(NULL) != SERIALIZER_OK)
        {
            (void)printf("Failed on serializer_init\r\n");
        }
        else
        {
            /* Setup IoTHub client configuration */
            iotHubClientHandle = IoTHubClient_CreateFromConnectionString(connectionString, AMQP_Protocol);
            srand((unsigned int)time(NULL));
            int avgWindSpeed = 10;

            // Turn on Log 
            bool trace = true;
            (void)IoTHubClient_SetOption(iotHubClientHandle, "logtrace", &trace);

            if (iotHubClientHandle == NULL)
            {
                (void)printf("Failed on IoTHubClient_Create\r\n");
            }
            else
            {
#ifdef MBED_BUILD_TIMESTAMP
                // For mbed add the certificate information
                if (IoTHubClient_SetOption(iotHubClientHandle, "TrustedCerts", certificates) != IOTHUB_CLIENT_OK)
                {
                    (void)printf("failure to set option \"TrustedCerts\"\r\n");
                }
#endif // MBED_BUILD_TIMESTAMP

                myWeather = CREATE_MODEL_INSTANCE(WeatherStation, ContosoAnemometer);
                if (myWeather == NULL)
                {
                    (void)printf("Failed on CREATE_MODEL_INSTANCE\r\n");
                }
                else
                {
                    unsigned char* destination;
                    size_t destinationSize;

                    if (IoTHubClient_SetMessageCallback(iotHubClientHandle, IoTHubMessage, myWeather) != IOTHUB_CLIENT_OK)
                    {
                        printf("unable to IoTHubClient_SetMessageCallback\r\n");
                    }
                    else
                    {
						myWeather->DeviceId = "Edison01";
						myWeather->nt1 = rand() % 50;
						myWeather->et1 = rand() % 50;
						myWeather->st1 = rand() % 50;
						myWeather->wt1 = rand() % 50;
						myWeather->nt2 = rand() % 50;
						myWeather->et2 = rand() % 50;
						myWeather->st2 = rand() % 50;
						myWeather->wt2 = rand() % 50;

                        if (SERIALIZE(&destination, &destinationSize, myWeather->DeviceId, myWeather->nt1, myWeather->et1, myWeather->st1, myWeather->wt1, myWeather->nt2, 
							myWeather->et2, myWeather->st2, myWeather->wt2
							) != IOT_AGENT_OK)
                        {
                            (void)printf("Failed to serialize\r\n");
                        }
                        else
                        {
                            sendMessage(iotHubClientHandle, destination, destinationSize);
                        }

                        /* wait for commands */
                        (void)getchar();
                    }
                    DESTROY_MODEL_INSTANCE(myWeather);
                }
                IoTHubClient_Destroy(iotHubClientHandle);
            }
            serializer_deinit();
        }
        platform_deinit();
    }
}

void SendMessage2(void)
{
	unsigned char* destination;
	size_t destinationSize;
	if (SERIALIZE(&destination, &destinationSize, myWeather->DeviceId, myWeather->nt1, myWeather->et1, myWeather->st1, myWeather->wt1, myWeather->nt2,
		myWeather->et2, myWeather->st2, myWeather->wt2
	) != IOT_AGENT_OK)
	{
		(void)printf("Failed to serialize\r\n");
	}
	else
	{
		sendMessage(iotHubClientHandle, destination, destinationSize);
	}
}
```
If you replace your version of simplesample_amqp.c with the above and build it in VS 2015 it should build and run without any problems.

When it first runs you should see it send the random values to Device Explorer.

Then as a test I sent the following commands to the device:
```
{"Name":"SetGreenTime", "Parameters":{"north_green_time":3,"north_priority":6,"east_green_time":4,"east_priority":8,"south_green_time":7,"south_priority":14,"west_green_time":9,"west_priority":18}}

{"Name":"SetGreenTime", "Parameters":{"north_green_time":36,"north_priority":66,"east_green_time":46,"east_priority":86,"south_green_time":76,"south_priority":146,"west_green_time":96,"west_priority":186}}
```
After you send either of the above commands, you should see the sample send back a message with the new values.  You should be able to toggle between the two as a verification that the values are changing and updating.  Based upon my limited testing I was able to change them back and forth a few times.

I hope this will serve as a solution to what you were trying to accomplish, or at least a better starting point to continue working from.

Thanks,

**Comment by cnvery** [comment]
Hi @jtanner-msft,

I tried your code on Windows Microsoft Visual Studio 2015, it worked perfectly. It can continuously send and receive messages.

However, it does not work on my Intel Edison with same code. It cant even establish connection with server via AMQP. So, not even able to send and receive first message.

Error as per below:
```
root@Edison1:~# ./Edison1
Info: IoT Hub SDK for C, version 1.0.5
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/iothub_client/src/iothubtransportamqp.c Func:IoTHubTransportAMQP_SetOption Line:1412 Invalid option (logtrace) passed to uAMQP transport SetOption()
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/iothub_client/src/iothub_client_ll.c Func:IoTHubClient_LL_SetOption Line:758 underlying transport failed, returned = IOTHUB_CLIENT_ERROR
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/iothub_client/src/iothub_client.c Func:IoTHubClient_SetOption Line:523 IoTHubClient_LL_SetOption failed
Mode: 1.
Junction 1 North Traffic is: 11
Junction 1 East Traffic is: 29
Junction 1 South Traffic is: 32
Junction 1 West Traffic is: 33
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/azure-c-shared-utility/adapters/socketio_berkeley.c Func:socketio_open Line:274 Failure: getaddrinfo failure -3.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/azure-c-shared-utility/src/tlsio_openssl.c Func:tlsio_openssl_open Line:526 Error in xio_open.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/iothub_client/src/iothubtransportamqp.c Func:establishConnection Line:456 Failed to open the connection with CBS.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/iothub_client/src/iothubtransportamqp.c Func:IoTHubTransportAMQP_DoWork Line:1222 AMQP transport failed to establish connection with service.
IoTHubClient accepted the message for delivery
Serialize Success!!
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/azure-c-shared-utility/adapters/socketio_berkeley.c Func:socketio_open Line:274 Failure: getaddrinfo failure -3.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/azure-c-shared-utility/src/tlsio_openssl.c Func:tlsio_openssl_open Line:526 Error in xio_open.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/iothub_client/src/iothubtransportamqp.c Func:establishConnection Line:456 Failed to open the connection with CBS.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/iothub_client/src/iothubtransportamqp.c Func:IoTHubTransportAMQP_DoWork Line:1222 AMQP transport failed to establish connection with service.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/azure-c-shared-utility/adapters/socketio_berkeley.c Func:socketio_open Line:274 Failure: getaddrinfo failure -3.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/azure-c-shared-utility/src/tlsio_openssl.c Func:tlsio_openssl_open Line:526 Error in xio_open.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/iothub_client/src/iothubtransportamqp.c Func:establishConnection Line:456 Failed to open the connection with CBS.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/iothub_client/src/iothubtransportamqp.c Func:IoTHubTransportAMQP_DoWork Line:1222 AMQP transport failed to establish connection with service.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/azure-c-shared-utility/adapters/socketio_berkeley.c Func:socketio_open Line:274 Failure: getaddrinfo failure -3.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/azure-c-shared-utility/src/tlsio_openssl.c Func:tlsio_openssl_open Line:526 Error in xio_open.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/iothub_client/src/iothubtransportamqp.c Func:establishConnection Line:456 Failed to open the connection with CBS.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/iothub_client/src/iothubtransportamqp.c Func:IoTHubTransportAMQP_DoWork Line:1222 AMQP transport failed to establish connection with service.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/azure-c-shared-utility/adapters/socketio_berkeley.c Func:socketio_open Line:274 Failure: getaddrinfo failure -3.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/azure-c-shared-utility/src/tlsio_openssl.c Func:tlsio_openssl_open Line:526 Error in xio_open.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/iothub_client/src/iothubtransportamqp.c Func:establishConnection Line:456 Failed to open the connection with CBS.
Error: Time:Tue Aug 30 05:52:03 2016 File:/home/osboxes/Source/azure-iot-sdks/c/iothub_client/src/iothubtransportamqp.c Func:IoTHubTransportAMQP_DoWork Line:1222 AMQP transport failed to establish connection with service.
```

My code as per below:
```
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <mraa.h>

// Receive JSON Messages Library
#include <string.h>

// Microsoft Azure IoT Library
#include "serializer.h"
#include "azure_c_shared_utility/threadapi.h"
#include "azure_c_shared_utility/sastoken.h"
#include "iothub_client.h"
#include "iothubtransportamqp.h"
#include "iothub_client_ll.h"
#include "azure_c_shared_utility/platform.h"

#ifdef MBED_BUILD_TIMESTAMP
#include "certs.h"
#endif // MBED_BUILD_TIMESTAMP

#define GRN   "\x1B[32m"
#define YEL   "\x1B[33m"
#define CYN   "\x1B[36m"
#define RESET "\x1B[0m"

/*String containing Hostname, Device Id & Device Key in the format:             */
/*  "HostName=<host_name>;DeviceId=<device_id>;SharedAccessKey=<device_key>"    */
static const char* connectionString = "HostName=trafficIoT.azure-devices.net;DeviceId=Edison01;SharedAccessKey=WxVdxNDST9AVqtr9W1g48CR9f3MZXWITs+hZjtIr+L0=";

const char certificates[] =
/* Baltimore */
"-----BEGIN CERTIFICATE-----\r\n"
"MIIDdzCCAl+gAwIBAgIEAgAAuTANBgkqhkiG9w0BAQUFADBaMQswCQYDVQQGEwJJ\r\n"
"RTESMBAGA1UEChMJQmFsdGltb3JlMRMwEQYDVQQLEwpDeWJlclRydXN0MSIwIAYD\r\n"
"VQQDExlCYWx0aW1vcmUgQ3liZXJUcnVzdCBSb290MB4XDTAwMDUxMjE4NDYwMFoX\r\n"
"DTI1MDUxMjIzNTkwMFowWjELMAkGA1UEBhMCSUUxEjAQBgNVBAoTCUJhbHRpbW9y\r\n"
"ZTETMBEGA1UECxMKQ3liZXJUcnVzdDEiMCAGA1UEAxMZQmFsdGltb3JlIEN5YmVy\r\n"
"VHJ1c3QgUm9vdDCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAKMEuyKr\r\n"
"mD1X6CZymrV51Cni4eiVgLGw41uOKymaZN+hXe2wCQVt2yguzmKiYv60iNoS6zjr\r\n"
"IZ3AQSsBUnuId9Mcj8e6uYi1agnnc+gRQKfRzMpijS3ljwumUNKoUMMo6vWrJYeK\r\n"
"mpYcqWe4PwzV9/lSEy/CG9VwcPCPwBLKBsua4dnKM3p31vjsufFoREJIE9LAwqSu\r\n"
"XmD+tqYF/LTdB1kC1FkYmGP1pWPgkAx9XbIGevOF6uvUA65ehD5f/xXtabz5OTZy\r\n"
"dc93Uk3zyZAsuT3lySNTPx8kmCFcB5kpvcY67Oduhjprl3RjM71oGDHweI12v/ye\r\n"
"jl0qhqdNkNwnGjkCAwEAAaNFMEMwHQYDVR0OBBYEFOWdWTCCR1jMrPoIVDaGezq1\r\n"
"BE3wMBIGA1UdEwEB/wQIMAYBAf8CAQMwDgYDVR0PAQH/BAQDAgEGMA0GCSqGSIb3\r\n"
"DQEBBQUAA4IBAQCFDF2O5G9RaEIFoN27TyclhAO992T9Ldcw46QQF+vaKSm2eT92\r\n"
"9hkTI7gQCvlYpNRhcL0EYWoSihfVCr3FvDB81ukMJY2GQE/szKN+OMY3EU/t3Wgx\r\n"
"jkzSswF07r51XgdIGn9w/xZchMB5hbgF/X++ZRGjD8ACtPhSNzkE1akxehi/oCr0\r\n"
"Epn3o0WC4zxe9Z2etciefC7IpJ5OCBRLbf1wbWsaY71k5h+3zvDyny67G7fyUIhz\r\n"
"ksLi4xaNmjICq44Y3ekQEe5+NauQrz4wlHrQMz2nZQ/1/I6eYs9HRCwBXbsdtTLS\r\n"
"R9I4LtD+gdwyah617jzV/OeBHRnDJELqYzmp\r\n"
"-----END CERTIFICATE-----\r\n"
/* MSIT */
"-----BEGIN CERTIFICATE-----\r\n"
"MIIFhjCCBG6gAwIBAgIEByeaqTANBgkqhkiG9w0BAQsFADBaMQswCQYDVQQGEwJJ\r\n"
"RTESMBAGA1UEChMJQmFsdGltb3JlMRMwEQYDVQQLEwpDeWJlclRydXN0MSIwIAYD\r\n"
"VQQDExlCYWx0aW1vcmUgQ3liZXJUcnVzdCBSb290MB4XDTEzMTIxOTIwMDczMloX\r\n"
"DTE3MTIxOTIwMDY1NVowgYsxCzAJBgNVBAYTAlVTMRMwEQYDVQQIEwpXYXNoaW5n\r\n"
"dG9uMRAwDgYDVQQHEwdSZWRtb25kMR4wHAYDVQQKExVNaWNyb3NvZnQgQ29ycG9y\r\n"
"YXRpb24xFTATBgNVBAsTDE1pY3Jvc29mdCBJVDEeMBwGA1UEAxMVTWljcm9zb2Z0\r\n"
"IElUIFNTTCBTSEEyMIICIjANBgkqhkiG9w0BAQEFAAOCAg8AMIICCgKCAgEA0eg3\r\n"
"p3aKcEsZ8CA3CSQ3f+r7eOYFumqtTicN/HJq2WwhxGQRlXMQClwle4hslAT9x9uu\r\n"
"e9xKCLM+FvHQrdswbdcaHlK1PfBHGQPifaa9VxM/VOo6o7F3/ELwY0lqkYAuMEnA\r\n"
"iusrr/466wddBvfp/YQOkb0JICnobl0JzhXT5+/bUOtE7xhXqwQdvDH593sqE8/R\r\n"
"PVGvG8W1e+ew/FO7mudj3kEztkckaV24Rqf/ravfT3p4JSchJjTKAm43UfDtWBpg\r\n"
"lPbEk9jdMCQl1xzrGZQ1XZOyrqopg3PEdFkFUmed2mdROQU6NuryHnYrFK7sPfkU\r\n"
"mYsHbrznDFberL6u23UykJ5jvXS/4ArK+DSWZ4TN0UI4eMeZtgzOtg/pG8v0Wb4R\r\n"
"DsssMsj6gylkeTyLS/AydGzzk7iWa11XWmjBzAx5ihne9UkCXgiAAYkMMs3S1pbV\r\n"
"S6Dz7L+r9H2zobl82k7X5besufIlXwHLjJaoKK7BM1r2PwiQ3Ov/OdgmyBKdHJqq\r\n"
"qcAWjobtZ1KWAH8Nkj092XA25epCbx+uleVbXfjQOsfU3neG0PyeTuLiuKloNwnE\r\n"
"OeOFuInzH263bR9KLxgJb95KAY8Uybem7qdjnzOkVHxCg2i4pd+/7LkaXRM72a1o\r\n"
"/SAKVZEhZPnXEwGgCF1ZiRtEr6SsxwUQ+kFKqPsCAwEAAaOCASAwggEcMBIGA1Ud\r\n"
"EwEB/wQIMAYBAf8CAQAwUwYDVR0gBEwwSjBIBgkrBgEEAbE+AQAwOzA5BggrBgEF\r\n"
"BQcCARYtaHR0cDovL2N5YmVydHJ1c3Qub21uaXJvb3QuY29tL3JlcG9zaXRvcnku\r\n"
"Y2ZtMA4GA1UdDwEB/wQEAwIBhjAdBgNVHSUEFjAUBggrBgEFBQcDAQYIKwYBBQUH\r\n"
"AwIwHwYDVR0jBBgwFoAU5Z1ZMIJHWMys+ghUNoZ7OrUETfAwQgYDVR0fBDswOTA3\r\n"
"oDWgM4YxaHR0cDovL2NkcDEucHVibGljLXRydXN0LmNvbS9DUkwvT21uaXJvb3Qy\r\n"
"MDI1LmNybDAdBgNVHQ4EFgQUUa8kJpz0aCJXgCYrO0ZiFXsezKUwDQYJKoZIhvcN\r\n"
"AQELBQADggEBAHaFxSMxH7Rz6qC8pe3fRUNqf2kgG4Cy+xzdqn+I0zFBNvf7+2ut\r\n"
"mIx4H50RZzrNS+yovJ0VGcQ7C6eTzuj8nVvoH8tWrnZDK8cTUXdBqGZMX6fR16p1\r\n"
"xRspTMn0baFeoYWTFsLLO6sUfUT92iUphir+YyDK0gvCNBW7r1t/iuCq7UWm6nnb\r\n"
"2DVmVEPeNzPR5ODNV8pxsH3pFndk6FmXudUu0bSR2ndx80oPSNI0mWCVN6wfAc0Q\r\n"
"negqpSDHUJuzbEl4K1iSZIm4lTaoNKrwQdKVWiRUl01uBcSVrcR6ozn7eQaKm6ZP\r\n"
"2SL6RE4288kPpjnngLJev7050UblVUfbvG4=\r\n"
"-----END CERTIFICATE-----\r\n"
/* *.azure-devices.net */
"-----BEGIN CERTIFICATE-----\r\n"
"MIIGcjCCBFqgAwIBAgITWgABtrNbz7vBeV0QWwABAAG2szANBgkqhkiG9w0BAQsF\r\n"
"ADCBizELMAkGA1UEBhMCVVMxEzARBgNVBAgTCldhc2hpbmd0b24xEDAOBgNVBAcT\r\n"
"B1JlZG1vbmQxHjAcBgNVBAoTFU1pY3Jvc29mdCBDb3Jwb3JhdGlvbjEVMBMGA1UE\r\n"
"CxMMTWljcm9zb2Z0IElUMR4wHAYDVQQDExVNaWNyb3NvZnQgSVQgU1NMIFNIQTIw\r\n"
"HhcNMTUwODI3MDMxODA0WhcNMTcwODI2MDMxODA0WjAeMRwwGgYDVQQDDBMqLmF6\r\n"
"dXJlLWRldmljZXMubmV0MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA\r\n"
"nXC/qBUdlnfIm5K3HYu0o/Mb5tNNcsr0xy4Do0Puwq2W1tz0ZHvIIS9VOANhkNCb\r\n"
"VyOncnP6dvmM/rYYKth/NQ8RUiZOYlROZ0SYC8cvxq9WOln4GXtEU8vNVqJbYrJj\r\n"
"rPMHfxqLzTE/0ZnQffnDT3iMUE9kFLHow0YgaSRU0KZsc9KAROmzBzu+QIB1WGKX\r\n"
"D7CN361tG1UuN68Bz7MSnbgk98Z+DjDxfusoDhiiy/Y9MLOJMt4WIy5BqL3lfLnn\r\n"
"r+JLqmpiFuyVUDacFQDprYJ1/AFgcsKYu/ydmASARPzqJhOGaC2sZP0U5oBOoBzI\r\n"
"bz4tfn8Bi0kJKmS53mQt+wIDAQABo4ICOTCCAjUwCwYDVR0PBAQDAgSwMB0GA1Ud\r\n"
"JQQWMBQGCCsGAQUFBwMBBggrBgEFBQcDAjAdBgNVHQ4EFgQUKpYehBSNA53Oxivn\r\n"
"aLCz3+eFUJ0wXQYDVR0RBFYwVIITKi5henVyZS1kZXZpY2VzLm5ldIIaKi5hbXFw\r\n"
"d3MuYXp1cmUtZGV2aWNlcy5uZXSCISouc3UubWFuYWdlbWVudC1henVyZS1kZXZp\r\n"
"Y2VzLm5ldDAfBgNVHSMEGDAWgBRRryQmnPRoIleAJis7RmIVex7MpTB9BgNVHR8E\r\n"
"djB0MHKgcKBuhjZodHRwOi8vbXNjcmwubWljcm9zb2Z0LmNvbS9wa2kvbXNjb3Jw\r\n"
"L2NybC9tc2l0d3d3Mi5jcmyGNGh0dHA6Ly9jcmwubWljcm9zb2Z0LmNvbS9wa2kv\r\n"
"bXNjb3JwL2NybC9tc2l0d3d3Mi5jcmwwcAYIKwYBBQUHAQEEZDBiMDwGCCsGAQUF\r\n"
"BzAChjBodHRwOi8vd3d3Lm1pY3Jvc29mdC5jb20vcGtpL21zY29ycC9tc2l0d3d3\r\n"
"Mi5jcnQwIgYIKwYBBQUHMAGGFmh0dHA6Ly9vY3NwLm1zb2NzcC5jb20wTgYDVR0g\r\n"
"BEcwRTBDBgkrBgEEAYI3KgEwNjA0BggrBgEFBQcCARYoaHR0cDovL3d3dy5taWNy\r\n"
"b3NvZnQuY29tL3BraS9tc2NvcnAvY3BzADAnBgkrBgEEAYI3FQoEGjAYMAoGCCsG\r\n"
"AQUFBwMBMAoGCCsGAQUFBwMCMA0GCSqGSIb3DQEBCwUAA4ICAQCrjzOSW+X6v+UC\r\n"
"u+JkYyuypXN14pPLcGFbknJWj6DAyFWXKC8ihIYdtf/szWIO7VooplSTZ05u/JYu\r\n"
"ZYh7fAw27qih9CLhhfncXi5yzjgLMlD0mlbORvMJR/nMl7Yh1ki9GyLnpOqMmO+E\r\n"
"yTpOiE07Uyt2uWelLHjMY8kwy2bSRXIp7/+A8qHRaIIdXNtAKIK5jo068BJpo77h\r\n"
"4PljCb9JFdEt6sAKKuaP86Y+8oRZ7YzU4TLDCiK8P8n/gQXH0vvhOE/O0n7gWPqB\r\n"
"n8KxsnRicop6tB6GZy32Stn8w0qktmQNXOGU+hp8OL6irULWZw/781po6d78nmwk\r\n"
"1IFl2TB4+jgyblvJdTM0rx8vPf3F2O2kgsRNs9M5qCI7m+he43Bhue0Fj/h3oIIo\r\n"
"Qx7X/uqc8j3VTNE9hf2A4wksSRgRydjAYoo+bduNagC5s7Eucb4mBG0MMk7HAQU9\r\n"
"m/gyaxqth6ygDLK58wojSV0i4RiU01qZkHzqIWv5FhhMjbFwyKEc6U35Ps7kP/1O\r\n"
"fdGm13ONaYqDl44RyFsLFFiiDYxZFDSsKM0WDxbl9ULAlVc3WR85kEBK6I+pSQj+\r\n"
"7/Z5z2zTz9qOFWgB15SegTbjSR7uk9mEVnj9KDlGtG8W1or0EGrrEDP2CMsp0oEj\r\n"
"VTJbZAxEaZ3cVCKva5sQUxFMjwG32g==\r\n"
"-----END CERTIFICATE-----\r\n";

// Define the Model
BEGIN_NAMESPACE(WeatherStation);

DECLARE_MODEL(ContosoAnemometer,
WITH_DATA(ascii_char_ptr, DeviceId),
WITH_DATA(int, mode),
WITH_DATA(int, nt1),
WITH_DATA(int, et1),
WITH_DATA(int, st1),
WITH_DATA(int, wt1),
WITH_ACTION(SetGreenTime, int, north_green_time, int, north_priority, int, east_green_time, int, east_priority, int, south_green_time, int, south_priority, int, west_green_time, int, west_priority)
);

END_NAMESPACE(WeatherStation);

// Global Variables
direction global_junction;
IOTHUB_CLIENT_HANDLE iotHubClientHandle = NULL;
ContosoAnemometer* myWeather = NULL;

DEFINE_ENUM_STRINGS(IOTHUB_CLIENT_CONFIRMATION_RESULT, IOTHUB_CLIENT_CONFIRMATION_RESULT_VALUES)

void sendCallback(IOTHUB_CLIENT_CONFIRMATION_RESULT result, void* userContextCallback)
{
    int messageTrackingId = (intptr_t)userContextCallback;

    (void)printf("Message Id: %d Received.\r\n", messageTrackingId);

    (void)printf("Result Call Back Called! Result is: %s \r\n", ENUM_TO_STRING(IOTHUB_CLIENT_CONFIRMATION_RESULT, result));
}

static void sendMessage(IOTHUB_CLIENT_HANDLE iotHubClientHandle, const unsigned char* buffer, size_t size)
{
    static unsigned int messageTrackingId;
    IOTHUB_MESSAGE_HANDLE messageHandle = IoTHubMessage_CreateFromByteArray(buffer, size);
    if (messageHandle == NULL)
    {
        printf("unable to create a new IoTHubMessage\r\n");
    }
    else
    {
        if (IoTHubClient_SendEventAsync(iotHubClientHandle, messageHandle, sendCallback, (void*)(uintptr_t)messageTrackingId) != IOTHUB_CLIENT_OK)
        {
            printf("failed to hand over the message to IoTHubClient");
        }
        else
        {
            printf("IoTHubClient accepted the message for delivery\r\n");
        }

        IoTHubMessage_Destroy(messageHandle);
    }
    free((void*)buffer);
    messageTrackingId++;
}

static IOTHUBMESSAGE_DISPOSITION_RESULT IoTHubMessage(IOTHUB_MESSAGE_HANDLE message, void* userContextCallback)
{
    IOTHUBMESSAGE_DISPOSITION_RESULT result;
    const unsigned char* buffer;
    size_t size;

    if (IoTHubMessage_GetByteArray(message, &buffer, &size) != IOTHUB_MESSAGE_OK)
    {
        printf("unable to IoTHubMessage_GetByteArray\r\n");
        result = EXECUTE_COMMAND_ERROR;
    }

    else
    {
        /*buffer is not zero terminated*/
        char* temp = malloc(size + 1);
        
	if (temp == NULL)
        {
            printf("failed to malloc\r\n");
            result = EXECUTE_COMMAND_ERROR;
        }
        
	else
        {
            memcpy(temp, buffer, size);
            temp[size] = '\0';
            EXECUTE_COMMAND_RESULT executeCommandResult = EXECUTE_COMMAND(userContextCallback, temp);
            result = (executeCommandResult == EXECUTE_COMMAND_ERROR) ? IOTHUBMESSAGE_ABANDONED : (executeCommandResult == EXECUTE_COMMAND_SUCCESS) ? IOTHUBMESSAGE_ACCEPTED : IOTHUBMESSAGE_REJECTED;
            free(temp);
	    (void)printf("Received Messages Success!!\r\n");
        }
    }
    return result;
}

void SendMessage1(void)
{
    if (platform_init() != 0)
    {
        printf("Failed to initialize the platform.\r\n");
    }

    else
    {
        if (serializer_init(NULL) != SERIALIZER_OK)
        {
            (void)printf("Failed on serializer_init\r\n");
        }

        else
        {
            /* Setup IoTHub client configuration */
	    iotHubClientHandle = IoTHubClient_CreateFromConnectionString(connectionString, AMQP_Protocol);
	    srand((unsigned int)time(NULL));

	    // Turn on Log 
	    bool trace = true;
	    (void)IoTHubClient_SetOption(iotHubClientHandle, "logtrace", &trace);

            if (iotHubClientHandle == NULL)
            {
                (void)printf("Failed on IoTHubClient_Create\r\n");
            }

            else
            {
// #ifdef MBED_BUILD_TIMESTAMP
                // For mbed add the certificate information
                if (IoTHubClient_SetOption(iotHubClientHandle, "TrustedCerts", certificates) != IOTHUB_CLIENT_OK)
                {
                    (void)printf("failure to set option \"TrustedCerts\"\r\n");
                }
// #endif // MBED_BUILD_TIMESTAMP

		myWeather = CREATE_MODEL_INSTANCE(WeatherStation, ContosoAnemometer);

                if (myWeather == NULL)
                {
                    (void)printf("Failed on CREATE_MODEL_INSTANCE\r\n");
                }

                else
                {
                    unsigned char* destination;
                    size_t destinationSize;

                    if (IoTHubClient_SetMessageCallback(iotHubClientHandle, IoTHubMessage, myWeather) != IOTHUB_CLIENT_OK)
                    {
                        printf("unable to IoTHubClient_SetMessageCallback\r\n");
                    }

                    else
                    {
                        myWeather->DeviceId = "Edison01";
			myWeather->mode = 1;
                        myWeather->nt1 = rand() % 50;
                        myWeather->et1 = rand() % 50;
                        myWeather->st1 = rand() % 50;
                        myWeather->wt1 = rand() % 50;

			(void)printf(CYN "Mode: %d.\r\n" RESET, myWeather->mode);
			(void)printf(CYN "Junction 1 North Traffic is: %d\r\n" RESET, myWeather->nt1);
			(void)printf(CYN "Junction 1 East Traffic is: %d\r\n" RESET, myWeather->et1);
			(void)printf(CYN "Junction 1 South Traffic is: %d\r\n" RESET, myWeather->st1);
			(void)printf(CYN "Junction 1 West Traffic is: %d\r\n" RESET, myWeather->wt1);

                        if (SERIALIZE(&destination, &destinationSize, myWeather->DeviceId, myWeather->mode, myWeather->nt1, myWeather->et1, myWeather->st1, myWeather->wt1) != IOT_AGENT_OK)
                        {
                            (void)printf("Failed to serialize\r\n");
                        }

                        else
                        {
                            sendMessage(iotHubClientHandle, destination, destinationSize);
                            (void)printf("Serialize Success!!\r\n");
                        }

                        /* wait for commands */
                        (void)getchar();
                    }
                    DESTROY_MODEL_INSTANCE(myWeather);
                }
                IoTHubClient_Destroy(iotHubClientHandle);
            }
            serializer_deinit();
        }
        platform_deinit();
    }
}

void SendMessage2(void)
{
	unsigned char* destination;
	size_t destinationSize;
	
	myWeather->mode = 1;
	myWeather->nt1 = rand() % 50;
	myWeather->et1 = rand() % 50;
	myWeather->st1 = rand() % 50;
	myWeather->wt1 = rand() % 50;

	delay_seconds(35);

	if (SERIALIZE(&destination, &destinationSize, myWeather->DeviceId, myWeather->mode, myWeather->nt1, myWeather->et1, myWeather->st1, myWeather->wt1) != IOT_AGENT_OK)
	{
		(void)printf("Failed to serialize\r\n");
	}
	else
	{
		sendMessage(iotHubClientHandle, destination, destinationSize);
	}
}

EXECUTE_COMMAND_RESULT SetGreenTime(ContosoAnemometer* device, int north_green_time, int north_priority, int east_green_time, int east_priority, int south_green_time, int south_priority, int west_green_time, int west_priority)
{
    (void)device;
    (void)printf(YEL "North Green Time is: %d\r\n" RESET, north_green_time);
    (void)printf(YEL "North Priority is: %d\r\n" RESET, north_priority);
    (void)printf(YEL "East Green Time is: %d\r\n" RESET, east_green_time);
    (void)printf(YEL "East Priority is: %d\r\n" RESET, east_priority);
    (void)printf(YEL "South Green Time is: %d\r\n" RESET, south_green_time);
    (void)printf(YEL "South Priority is: %d\r\n" RESET, south_priority);
    (void)printf(YEL "West Green Time is: %d\r\n" RESET, west_green_time);
    (void)printf(YEL "West Priority is: %d\r\n" RESET, west_priority);

    global_junction.north_green_time = north_green_time;
    global_junction.north_priority = north_priority;
    global_junction.east_green_time = east_green_time;
    global_junction.east_priority = east_priority;
    global_junction.south_green_time = south_green_time;
    global_junction.south_priority = south_priority;
    global_junction.west_green_time = west_green_time;
    global_junction.west_priority = west_priority;

    SendMessage2();

    return EXECUTE_COMMAND_SUCCESS;
}
```

**Comment by cnvery** [comment]
Hi @jtanner-msft,

I found out that other users also faces similar AMQP problem just like me.

The link is as per below:
[simplesample_amqp execution failure on Debian Beagleboneblack](https://github.com/Azure/azure-iot-sdks/issues/290)
[[dmpreview][C][Linux] Can't get iothub_client_sample_amqp connect to iot hub](https://github.com/Azure/azure-iot-sdks/issues/611)

Both of them able to solve the problem by using non firewall blocked network.

Their solution is not working for my side.

Do you have any solution on this?

Thanks!

**Comment by jtanner-msft** [comment]
@cnvery,

  What you have done looks to be identical to what I had done to get the AMQP working correctly for myself on the Edison.  I'll try to run my modified code there and see if I experience the same problems you are seeing.  If I don't, then I would tend to think it might be something more specific to your local network like a firewall issue or something.

  If you make the certificate changes to just the simplesample_amqp.c does everything work correctly for you on the Edison?

**Comment by cnvery** [comment]
Hi @jtanner-msft,

I have tried simplesample_amqp.c with and without certicates, however the output still same.

**Comment by jtanner-msft** [comment]
Hi @cnvery,

  I’ve found that if you leave the certificates in place and simply remove the call to SendMessage2(), within the function SetGreenTime() then the sample appears to continue to work reliably in that I can continue to receive new messages that are sent from Cloud 2 Device.  You can continue to see that the new values are getting printed out whenever you send them, even if the client isn’t in turn sending them back at that same moment.  As I have time I’ll try to continue to investigate why the call to SendMessage2() causes the sample to hang when it receives the first Cloud 2 Device message, but when I looked under the debugger earlier I wasn’t able to see any obvious reason.  
  Either way I hope with that change you’ll be able to move forward with your testing and development for the time being.

Thanks,

**Comment by cnvery** [comment]
Hi @jtanner-msft,

I have tried again just like what you suggested.

However, the AMQP connection problem still occured.

You can refer to the error from this image.

![34234](https://cloud.githubusercontent.com/assets/15714095/18223582/081d8f26-71ee-11e6-9889-f37dd6a19807.PNG)

Now, I am unable to continue development for my project due to this problem.

**Comment by jtanner-msft** [comment]
Hi @cnvery,

  In looking up error code -3 for getaddrinfo it means:

# define EAI_AGAIN        -3    /* Temporary failure in name resolution.  */

 So it sounds like your Edison is not able to resolve the name of your IoTHub.  So it sounds like you might well have a firewall issue preventing connectivity from your Edison. 

 I would suggest that you test that you can actually access your IoT Hub from your device.  To do that you can use the following command:

  telnet <yourhubname>.azure-devices.net 5671

  If you don’t see any error messages then it would mean you’ve been able to establish a connection.  You can test the behavior from your desktop as well to confirm the behavior.

**Comment by jtanner-msft** [comment]
Hi @cnvery , I'm hoping that you were able to determine what you needed to do to unblock those ports on your Edison.  Since I haven't seen a response in over a week now I'm going to close this issue.  If you do need assistance in the future please feel free to either re-open this one or create a new issue that references this one.
